packages/core/src/mixer.ts

```diff
@@ -25,42 +23,129 @@ const comparator = (a: HeapElement, b: HeapElement): number => {
  */
 export class Mixer<Chain extends HoprCoreConnector> {
   private queue: Heap<HeapElement>
+  private timeout?: NodeJS.Timeout
+  private defer: DeferredPromise<void>
+  private deferEnd: DeferredPromise<void>
+  private endPromise: Promise<void>
+  private _done: boolean
 
   public WAIT_TIME = MAX_PACKET_DELAY
 
   constructor(private incrementer = Date.now) {
     this.queue = new Heap(comparator)
-  }
 
-  push(p: Packet<Chain>) {
-    this.queue.push([this.getPriority(), p])
+    this.defer = Defer<void>()
+    this.deferEnd = Defer<void>()
+
+    this._done = false
+
+    this.endPromise = this.deferEnd.promise.then(() => {
+      this._done = true
+    })
   }
 
-  // Can we pop an element?.
-  poppable(): boolean {
-    log(`Mixer has ${this.queue.length} elements`)
-    if (!this.queue.length) {
-      return false
+  /**
+   * Adds a packet to the mixer
+   * @param p
+   */
+  public push(p: Packet<Chain>) {
+    if (this._done) {
+      throw Error(`Mixer has ended. Could not accept any further messages.`)
     }
-    return this.queue.peek()[0] < this.due()
-  }
+    const newPriority = this.getPriority()
+    const topPriority = this.queue.length > 0 ? this.queue.peek()[0] : Number.MAX_SAFE_INTEGER
 
-  pop(): Packet<Chain> {
-    if (!this.poppable()) {
-      throw new Error('No packet is ready to be popped from mixer')
+    if (newPriority < topPriority) {
+      this.resetTimeout(newPriority)
     }
-    return this.queue.pop()[1]
+
+    log(`Added 1 packet to the mixer`)
+    this.queue.push([newPriority, p])
+  }
+
+  /**
+   * Get the number of packets that are currently
+   * in the mixer
+   */
+  public get length() {
+    return this.queue.length
+  }
+
+  /**
+   * Adjust the internal timeout with latest priority
+   * @param newPriority
+   */
+  private resetTimeout(newPriority: number) {
+    clearTimeout(this.timeout)
+    this.timeout = setTimeout(() => {
+      this.defer.resolve()
+    }, newPriority - Date.now())
   }
 
   notEmpty(): boolean {
     return this.queue.length > 0
   }
 
-  private due(): number {
-    return this.incrementer()
+  /**
+   * Ends the mixer.
+   */
+  public end(): void {
+    log(`Ending mixer. Mixer will not accept any further messages`)
+
+    if (this._done) {
+      return
+    }
+
+    this.deferEnd.resolve()
+  }
+
+  public get done() {
```

Comment:
Why do we need a done method?

Reply:
... because we need to distinguish between: (1) the mixer has no elements - waiting (2) the mixer might have elements, but it has stopped processing new packets

Reply:
We don't need the second.

Reply:
removed `done` and `length`